patches/server/1036-Add-predicate-for-blocks-when-raytracing.patch

```diff
@@ -0,0 +1,120 @@
++        // Paper Start
++        return rayTraceEntities((io.papermc.paper.math.Position) start, direction, maxDistance, raySize, filter);
++    }
++
++    public RayTraceResult rayTraceEntities(io.papermc.paper.math.Position start, Vector direction, double maxDistance, double raySize, Predicate<Entity> filter) {
+         Preconditions.checkArgument(start != null, "Location start cannot be null");
+-        Preconditions.checkArgument(this.equals(start.getWorld()), "Location start cannot be in a different world");
+-        start.checkFinite();
++        Preconditions.checkArgument(!(start instanceof Location location) || this.equals(location.getWorld()), "Location start cannot be in a different world");
++        org.bukkit.util.NumberConversions.checkFinite(start.x(), "x not finite");
++        org.bukkit.util.NumberConversions.checkFinite(start.y(), "y not finite");
++        org.bukkit.util.NumberConversions.checkFinite(start.z(), "z not finite");
++        // Paper End
+ 
+         Preconditions.checkArgument(direction != null, "Vector direction cannot be null");
+         direction.checkFinite();
+@@ -1169,9 +1177,18 @@ public class CraftWorld extends CraftRegionAccessor implements World {
+ 
+     @Override
+     public RayTraceResult rayTraceBlocks(Location start, Vector direction, double maxDistance, FluidCollisionMode fluidCollisionMode, boolean ignorePassableBlocks) {
++        // Paper start
++        return this.rayTraceBlocks(start, direction, maxDistance, fluidCollisionMode, ignorePassableBlocks, null);
++    }
++
++    @Override
++    public RayTraceResult rayTraceBlocks(io.papermc.paper.math.Position start, Vector direction, double maxDistance, FluidCollisionMode fluidCollisionMode, boolean ignorePassableBlocks, Predicate<Block> passThrough) {
+         Preconditions.checkArgument(start != null, "Location start cannot be null");
+-        Preconditions.checkArgument(this.equals(start.getWorld()), "Location start cannot be in a different world");
+-        start.checkFinite();
++        Preconditions.checkArgument(!(start instanceof Location location) || this.equals(location.getWorld()), "Location start cannot be in a different world");
++        org.bukkit.util.NumberConversions.checkFinite(start.x(), "x not finite");
++        org.bukkit.util.NumberConversions.checkFinite(start.y(), "y not finite");
++        org.bukkit.util.NumberConversions.checkFinite(start.z(), "z not finite");
++        // Paper end
+ 
+         Preconditions.checkArgument(direction != null, "Vector direction cannot be null");
+         direction.checkFinite();
+@@ -1184,16 +1201,23 @@ public class CraftWorld extends CraftRegionAccessor implements World {
+         }
+ 
+         Vector dir = direction.clone().normalize().multiply(maxDistance);
+-        Vec3 startPos = CraftLocation.toVec3D(start);
++        Vec3 startPos = new Vec3(start.x(), start.y(), start.z()); // Paper
```

Comment:
This should probably live in the MCUtil patch as a utility.

Reply:
Added.